### PR TITLE
fix(chain): verify orphan approvals

### PR DESCRIPTION
**Pull request by mikhailOK**
Partially verify approvals for orphans.

Protocol changes:
- `Vec<Option<Approvals>>` in block header now has elements for slashed validators (must be None)
- Doomslug threshold from 2/3 of non-slashed to 2/3 of all stake
- `BlockHeaderV3` with `prev_height` field

Test plan
---------
unit test

### CODE REVIEW

File: chain/chain/src/chain.rs [hunk 1]
```diff
@@ -3258,6 +3267,35 @@ impl<'a> ChainUpdate<'a> {
...
+        Ok(())
+    }
+
+    #[cfg(feature = "protocol_feature_block_header_v3")]
+    fn verify_orphan_header_approvals(&mut self, header: &BlockHeader) -> Result<(), Error> {
+        let prev_hash = header.prev_hash();
+        let prev_height = match header.prev_height() {
+            None => {
+                // this will accept orphans of V1 and V2
+                // TODO: reject header V1 and V2 after a certain height
```
Comment (bowenwang1996) [inline chain/chain/src/chain.rs:3281]: don't think we need to do this. We currently reject blocks that are older than tail right away, so after the network has been upgraded it wouldn't be long before the tail becomes V3

File: chain/chain/src/doomslug.rs [hunk 2]
```diff
@@ -420,25 +426,27 @@ impl Doomslug {
...
-        let threshold1 = stakes.iter().map(|(x, _)| x).sum::<Balance>() * 2 / 3;
-        let threshold2 = stakes.iter().map(|(_, x)| x).sum::<Balance>() * 2 / 3;
+        let threshold1 = stakes.iter().map(|(x, _, _)| x).sum::<Balance>() * 2 / 3;
+        let threshold2 = stakes.iter().map(|(_, x, _)| x).sum::<Balance>() * 2 / 3;
 
         let approved_stake1 = approvals
             .iter()
             .zip(stakes.iter())
-            .map(|(approval, (stake, _))| if approval.is_some() { *stake } else { 0 })
+            .filter(|(_, (_, _, is_slashed))| !*is_slashed)
```
Comment (bowenwang1996) [inline chain/chain/src/doomslug.rs:441]: might be better to use `filter_map` directly https://nnethercote.github.io/perf-book/iterators.html#chaining

File: chain/epoch_manager/src/lib.rs [hunk 3]
```diff
@@ -666,10 +666,30 @@ impl EpochManager {
...
 
+    /// get_heuristic_block_approvers_ordered: block producers for epoch
+    /// get_all_block_approvers_ordered: block producers for epoch, slashing info, sometimes block producers for next epoch
+    pub fn get_heuristic_block_approvers_ordered(
+        &mut self,
+        epoch_id: &EpochId,
+    ) -> Result<Vec<ApprovalStake>, EpochError> {
+        let epoch_info = self.get_epoch_info(epoch_id)?;
+        let mut settlement = vec![];
+        for validator_id in epoch_info.block_producers_settlement.iter() {
```
Comment (bowenwang1996) [inline chain/epoch_manager/src/lib.rs:677]: there is an existing function `get_all_block_producers_settlement`

File: core/primitives/src/block_header.rs [hunk 4]
```diff
@@ -287,6 +360,14 @@ impl BlockHeader {
             next_bp_hash,
             block_merkle_root,
         };
+        #[cfg(not(feature = "protocol_feature_block_header_v3"))]
+        let last_header_v2_version = 9999;
```
Comment (bowenwang1996) [inline core/primitives/src/block_header.rs:364]: can we use an option to avoid this?

File: neard/src/runtime.rs [hunk 5]
```diff
@@ -770,6 +773,47 @@ impl RuntimeAdapter for NightshadeRuntime {
...
+        &self,
+        epoch_id: &EpochId,
+        doomslug_threshold_mode: DoomslugThresholdMode,
+        prev_block_hash: &CryptoHash,
+        prev_block_height: BlockHeight,
+        block_height: BlockHeight,
+        approvals: &[Option<Signature>],
+    ) -> Result<(), Error> {
+        let mut epoch_manager = self.epoch_manager.as_ref().write().expect(POISONED_LOCK_ERR);
+        let info =
```
Comment (bowenwang1996) [inline neard/src/runtime.rs:787]: let's move this two lines in two a block so that the lock is dropped immediately after we get the necessary data.

**Review by bowenwang1996 — COMMENTED** [review]
Verifying all signatures in a header isn't very cheap. We should make sure that we only do it once for every block, i.e, if a block is orphaned and later processed, we should not verify signature again.

**Comment by bowenwang1996** [comment]
@SkidanovAlex Will this change break the bridge?

File: chain/chain/src/chain.rs (hunk 1 above)
Comment (mikhailOK) [reply chain/chain/src/chain.rs:3281]: Where do we reject them?
Seems like some checks are missing

File: chain/epoch_manager/src/lib.rs (hunk 3 above)
Comment (mikhailOK) [reply chain/epoch_manager/src/lib.rs:677]: `get_all_block_producers_settlement` takes a block hash, which is not available

File: chain/chain/src/chain.rs (hunk 1 above)
Comment (bowenwang1996) [reply chain/chain/src/chain.rs:3281]: in `ClientActor::receive_block`, you might want to merge with master

Comment (mikhailOK) [reply chain/chain/src/chain.rs:3281]: can't see the check, made `test_header_version_downgrade` with `should_panic`

File: chain/epoch_manager/src/lib.rs (hunk 3 above)
Comment (bowenwang1996) [reply chain/epoch_manager/src/lib.rs:679]: I see, but we can use the `epoch_validators_ordered` cache

File: chain/chain/src/chain.rs (hunk 1 above)
Comment (bowenwang1996) [reply chain/chain/src/chain.rs:3281]: https://github.com/near/nearcore/blob/4a9ecaed34a9bce3ba7c469af48a78d9d02e8461/chain/client/src/client_actor.rs#L960

Comment (mikhailOK) [reply chain/chain/src/chain.rs:3281]: We reject old heights, but don't reject old versions - a malicious node can produce V1 at any height, and it will be approved

Comment (bowenwang1996) [reply chain/chain/src/chain.rs:3281]: oh I misunderstood what you meant

**Comment by frol** [comment]
It seems to be in conflict with #3804 which also introduced BlockHeader v3

**Comment by frol** [comment]
What is the state of this PR?

**Review by bowenwang1996 — COMMENTED** [review]
@mikhailOK you need to bump nightly protocol version again for this change so that it automatically triggers a hard fork when we deploy it to betanet.

File: chain/client/tests/process_blocks.rs [hunk 6]
```diff
@@ -2408,6 +2410,51 @@ fn test_not_broadcast_block_on_accept() {
     assert!(network_adapter.requests.read().unwrap().is_empty());
 }
 
+#[test]
+#[should_panic]
+// TODO: reject header version downgrade
```
Comment (bowenwang1996) [inline chain/client/tests/process_blocks.rs:2415]: are you going to do this in a separate PR? If so, it is better to add the issue number here

**Review by bowenwang1996 — APPROVED** [review]
